### PR TITLE
postgresql16*: new ports

### PR DESCRIPTION
**Pull request by dgilman**
#### Description

The most important change here is that I have changed ownership from @jyrkiwahlstedt and made the port openmaintainer. He has been busy elsewhere for a while and hasn't been on GitHub in some time so I think the change is justified. It is also a new port anyway.

This is a beta release, when the official one comes out I'd rename the port to postgresql16.

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [x] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

**Comment by reneeotten** [comment]
can we just wait for the official release - any idea on when that will be?

**Comment by dgilman** [comment]
It will be released some time in the next month or two. I think there have been beta versions in MacPorts before, though, because the portfile had a livecheck regex commented out that only matches pre-release versions.

**Comment by reneeotten** [comment]
> It will be released some time in the next month or two. I think there have been beta versions in MacPorts before, though, because the portfile had a livecheck regex commented out that only matches pre-release versions.

I am not sure, not seeing any previous `-devel` ports with a quick look, but who knows. I personally don't really care about it, but regardless the CI fails for the "-doc" port so that should be fixed. Then we can see what others think about it.

**Comment by dgilman** [comment]
It looks like Jyrki went from `\d\dbeta\d` to `\d\d[.]\d` without additional -dev ports. So that option is also on the table. https://github.com/macports/macports-ports/commit/f4a7e25cdd9fa4c32d339cdb96fa8cfea27f9b02

**Comment by barracuda156** [comment]
@dgilman @reneeotten Please allow me to test prior to merging.

**Comment by reneeotten** [comment]
> It looks like Jyrki went from `\d\dbeta\d` to `\d\d[.]\d` without additional -dev ports. So that option is also on the table. https://github.com/macports/macports-ports/commit/f4a7e25cdd9fa4c32d339cdb96fa8cfea27f9b02 

Yeah, that would avoid adding a new port and doing the "replaced_by" stuff as just renaming a port cannot be done. Only drawback in that approach is likely that one would need to increase the "epoch" as I think that version "16" would not be considered newer than, for example, "16rc2". But that should be tested to make sure.

**Comment by barracuda156** [comment]
@dgilman @reneeotten To begin with, we need `legacysupport`:
```
In file included from ../../src/include/pgstat.h:15,
                 from controldata_utils.c:38:
../../src/include/portability/instr_time.h: In function ‘pg_clock_gettime_ns’:
../../src/include/portability/instr_time.h:116: warning: implicit declaration of function ‘clock_gettime’
../../src/include/portability/instr_time.h:116: error: ‘CLOCK_REALTIME’ undeclared (first use in this function)
../../src/include/portability/instr_time.h:116: error: (Each undeclared identifier is reported only once
../../src/include/portability/instr_time.h:116: error: for each function it appears in.)
```

**Comment by barracuda156** [comment]
Then, something is broken here:
```
/usr/bin/gcc-4.2 -std=gnu99 -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -pipe -Os -arch ppc -I../../../src/include  -isystem/opt/local/include/LegacySupport -I/opt/local/include -I/opt/local/include/libxml2  -I/opt/local/include  -c -o vacuum.o vacuum.c
/usr/bin/gcc-4.2 -std=gnu99 -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -pipe -Os -arch ppc -I../../../src/include  -isystem/opt/local/include/LegacySupport -I/opt/local/include -I/opt/local/include/libxml2  -I/opt/local/include  -c -o vacuumparallel.o vacuumparallel.c
{standard input}:605:Parameter syntax error (parameter 3)
{standard input}:615:Parameter syntax error (parameter 3)
{standard input}:1030:Parameter syntax error (parameter 3)
gnumake[3]: *** [vacuumparallel.o] Error 1
gnumake[3]: *** Waiting for unfinished jobs....
{standard input}:185:Parameter syntax error (parameter 3)
gnumake[3]: *** [vacuum.o] Error 1
```
Let me look into this.

**Comment by barracuda156** [comment]
I think I have found broken code. Let me make a patch and try.

**Comment by barracuda156** [comment]
@dgilman @reneeotten Turned out, it was broken on purpose in https://github.com/postgres/postgres/commit/8ded65682bee2a1c04392a88e0df0f4fc7552623

Anyway, this patch fixes the build (`legacysupport` PG still needed, since upstream has removed the check for `clock_gettime`):
[patch-fix-ppc-asm.diff.zip](https://github.com/macports/macports-ports/files/12465814/patch-fix-ppc-asm.diff.zip)

**Comment by dgilman** [comment]
The commit message mentions that modern assemblers can emit compatible instructions. Is it possible to blacklist old compilers instead of having to try and carry this patch forever?

**Comment by barracuda156** [comment]
@dgilman The problem here is that even if the comment is accurate (which we cannot be sure of), we have no selection of assemblers for macOS PowerPC. Using the latest GCC won’t help, it does not provide `as`.

From the looks of things upstream did a wrong thing removing the code.

It is possible that the code can be improved and made more concise. _Maybe_ `lwsync` can be supported via ppc intrinsics (just a guess here). But I can’t deal with this in a couple of hours.

Be sure, I do not want to carry a huge patch _forever_ myself (given that pretty likely I will have to test it and fix with subsequent updates). Solutions – either bring this to attention of upstream that their assumptions were wrong and consequently the move was wrong either, and should be reverted; or rewrite the code to make it more portable along releases.
Writing a brand new assembler for old PPC, I am afraid, is out of scope of being feasible.

**Comment by barracuda156** [comment]
Before I bothered to check commit history and find the breakage, I have already figured out to replace `lwsync` with either `isync` or `sync`. What I missed is that the new code uses invalid form of `lwarx` which smuggles in an extra argument.

As can be seen, the instruction takes 3 arguments: https://opensource.apple.com/source/xnu/xnu-792.6.76/osfmk/ppc/commpage/atomic.s.auto.html
While the code now passes 4. Of course this cannot work.

I suggest we merge what we have for now (just restoring a working upstream code), and later I will see if a) we can make `lwsync` work (it is mentioned in intrinsic header) and what will be a better way to fix broken `lwarx`/`ldarx`.

**Comment by barracuda156** [comment]
For the `legacysupport`, it is worth specifying the latest Darwin version which needs it. Check the portgroup. (It is not in the portfile now, or did I just overlook it?)

File: databases/postgresql16-dev/Portfile
Comment (ryandesign) [inline databases/postgresql16-dev/Portfile:22]: This is not aligned properly.

File: databases/postgresql16-dev/Portfile
Comment (ryandesign) [inline databases/postgresql16-dev/Portfile:83]: This is in the default ldflags; you should need to add it again.

File: databases/postgresql16-dev/Portfile
Comment (ryandesign) [inline databases/postgresql16-dev/Portfile:94]: This is ancient compatibility code for old versions of MacPorts base that don't have `universal_possible`.

File: databases/postgresql16-dev/Portfile
Comment (ryandesign) [inline databases/postgresql16-dev/Portfile:115]: Don't use `system` when a native Tcl equivalent exists.

File: databases/postgresql16-dev/Portfile
Comment (ryandesign) [inline databases/postgresql16-dev/Portfile:142]: It doesn't seem like we need two variants that do the same thing. Presumably the variant used to be called "python" and was then changed to "python3" and the python variant was retained to assist users upgrading from the previous variant name. No such assistance is needed here since this is a new port.

File: databases/postgresql16-server/Portfile
Comment (ryandesign) [inline databases/postgresql16-server/Portfile:49]: Don't use `system` when a native Tcl equivalent exists.

**Review by ryandesign — COMMENTED** [review]
In my view, there are two separate port naming strategies that can be employed:

1. foo and foo-devel. In this case, foo is the latest stable version, foo-devel is the latest development version, and they intentionally install files to the same locations and conflict with one another and the devel port is a drop-in replacement for the non-devel port. Examples: libpixman, cairo, pango, glib2, graphviz, minivmac.
2. bar1, bar2, bar3. In this case, each port provides the latest version of the specified major version of the software, they intentionally install files to different locations and do not conflict with one another. There is also a bar_select port to let you select which version of bar you want to use at the command line with the standard name without having to specify the version number. Examples: clang and llvm, gcc, php, perl5*, python**.

*perl5 is unique in that it [does not have a working perl_select port](https://trac.macports.org/ticket/29763). It predates the existence of `port select` so it's not a good model for what should be done today.

**python is unique in that Josh has chosen to use the -devel suffix on development versions of python for which there is not yet any stable version, such as python312-devel today. When the stable version comes around, Josh will copy the python312-devel port to python312, making adjustments along the way, and hopefully leaving the stub python312-devel port marked `replaced_by python312` for awhile. This is a lot of extra work that I view as unnecessary for this case, hence the manner in which I handle the php ports, for example php83 today is 8.3.0beta3 and will eventually become 8.3.0 final (with the corresponding necessary epoch increase).

If you wish to use the devel naming scheme for postgresql16, then the suffix must be `-devel` not `-dev`.

Many of the comments I made below may apply to previous postgresql ports as well. All of the postgresql ports should be kept as similar as possible. This has, to a great degree, not happened with these ports so far, so straightening that out and figuring out which differences between them are intentional and which are not and should be back- or forward-ported to the other ones may be a lot of work.

**Comment by dgilman** [comment]
Done:
- added barracuda as maintainer per comment above
- @barracuda156 are https://trac.macports.org/ticket/66689 and https://trac.macports.org/ticket/67203 done? if so, i will close here. it seems like we are ICU everywhere and PPC is building on your machine
- update to version 16rc1
- switched port name to just `postgresql16`.  In the past the port (without suffix) went from beta -> rc -> final release, and that's probably what the end user would expect from a postgresql16 port anyway.
- addressed review feedback
- added many Closes: to commit messages

User homedir / shell changes:
There are a few issues ( https://trac.macports.org/ticket/24180 https://trac.macports.org/ticket/14083 https://trac.macports.org/ticket/64286 ) that have to do with a poor initial choice of shell and homedir for the postgresql user in the `-server` port. The `add_users` Tcl function doesn't really support migrating account properties after a user account has been set up. I tried throwing `system "dscl ..."` into the Portfile, it fails because of some sandboxing (?) that denies dscl access to change the properties.

Barring some trick to get around the sandbox that I don't know about, I think the way forward here is to just set the defaults right in postgresql16 and let the fixes trickle out over time. They aren't breaking any fundamental functionality with the port.

Backporting changes:
I didn't do any backporting of changes to earlier PG releases in this PR. However, I am thinking the way forward is to mark 11 and before with the depreciated portgroup and only copy changes to 12+. 11 goes out of support in November.

TODO:
- backport fixes to old versions
- kill `-doc` portfile in favor of moving everything into the select portgroup
- https://trac.macports.org/ticket/66529 and https://trac.macports.org/ticket/67638

**Comment by barracuda156** [comment]
@dgilman `postgresql13` and `postgresql14` should build fine now on PPC. I have just re-built `postgresql13` in Rosetta to make sure. So yes, those tickets can be closed, I believe.
(Not sure if the fix has been backported to all earlier versions where it may be needed though.)

Re sandboxing, while I do not know whether it is relevant, but we use the following in some ports: https://github.com/macports/macports-ports/blob/ffe6352713a89b607460d5140d1952579e446c0d/science/madness/Portfile#L129-L132

**Comment by mascguy** [comment]
> Only drawback in that approach is likely that one would need to increase the "epoch" as I think that version "16" would not be considered newer than, for example, "16rc2". But that should be tested to make sure.

To avoid dealing with all of that, the current version of the port could start with 15. So something like `15.99.rc1`.

**Comment by dgilman** [comment]
I've uncommented the dscl stuff. Not sure what I was running into before but it is now running correctly / offering the correct privs. I think it may have been an issue of the permission popup modal the system does not correctly firing or me accidentally denying it.

This PR is good for final review

**Review by mascguy — APPROVED** [review]

**Comment by ryandesign** [comment]
> > Only drawback in that approach is likely that one would need to increase the "epoch" as I think that version "16" would not be considered newer than, for example, "16rc2". But that should be tested to make sure.
> 
> To avoid dealing with all of that, the current version of the port could start with 15. So something like `15.99.rc1`.

The epoch exists for cases like this. It's much better to use the epoch as intended rather than lie to the user about the port's version number.

**Pull request merged**
